File: packages/gbnf/python/gbnf/grammar_graph/graph.py
from __future__ import annotations

from collections.abc import Iterable
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .grammar_graph_types import (
        Pointers,
        ResolvedGraphPointer,
        UnresolvedRule,
        ValidInput,
    )


from ..utils.errors import InputParseError
from ..utils.is_point_in_range import is_point_in_range
from .colorize import colorize
from .get_input_as_code_points import get_input_as_code_points
from .get_serialized_rule_key import get_serialized_rule_key
from .graph_node import GraphNode
from .graph_pointer import GraphPointer
from .rule_ref import RuleRef
from .type_guards import (
    is_range,
    is_rule_char,
    is_rule_char_exclude,
    is_rule_end,
    is_rule_ref,
)

RootNode = dict[int, GraphNode]


def make_pointers() -> Pointers:
    """
    Create a set of pointers with a custom comparison based on pointer ID.
    """
    return set()


class Graph:
    roots: dict[int, RootNode]
    grammar: str
    __rootNode__: RootNode | None
    previous_code_points: list[int]

    def __init__(
        self,
        grammar: str,
        stackedRules: list[list[list[UnresolvedRule]]],
        rootId: int,
    ):
        self.roots = {}
        self.grammar = grammar
        self.previous_code_points = []
        rule_refs: list[RuleRef] = []
        unique_rules: dict[str, UnresolvedRule] = {}

        for stackId in range(len(stackedRules)):
            stack = stackedRules[stackId]
            nodes: dict[int, GraphNode] = {}
            for pathId in range(len(stack)):
                path = stack[pathId]
                node: GraphNode | None = None
                for stepId in range(len(path) - 1, -1, -1):
                    next_node: GraphNode | None = node
                    rule = stack[pathId][stepId]
                    unique_rules[get_serialized_rule_key(rule)] = rule
                    if is_rule_ref(rule):
                        rule_refs.append(rule)
                    # rules coming in may be identical but have different references.
                    # here, we ensure we always use the same reference for an identical rule.
                    # this makes future comparisons easier.
                    unique_rule = unique_rules.get(get_serialized_rule_key(rule))
                    if unique_rule is None:
                        raise ValueError("Could not get unique rule")
                    node = GraphNode(
                        unique_rule,
                        {
                            "stackId": stackId,
                            "pathId": pathId,
                            "stepId": stepId,
                        },
                        next_node,
                    )

                if node is None:
                    raise ValueError("Could not get node")
                nodes[pathId] = node
            self.roots[stackId] = nodes

        self.root_node = self.roots.get(rootId)

        for rule_ref in rule_refs:
            referenced_nodes = set()
            referenced_nodes.update(self.get_root_node(rule_ref.value).values())
            rule_ref.nodes = referenced_nodes

    def get_root_node(self, value: int) -> RootNode:
        root_node = self.roots.get(value)
        if root_node is None:
            raise ValueError(f"Root node not found for value: {value}")
        return root_node

    @property
    def root_node(self) -> RootNode:
        if self.__rootNode__ is None:
            raise ValueError("Root node is not defined")
        return self.__rootNode__

    @root_node.setter
    def root_node(self, rootNode: RootNode | None):
        if rootNode is None:
            raise ValueError("Root node is not defined")
        self.__rootNode__ = rootNode

    def get_initial_pointers(self) -> Pointers:
        pointers = make_pointers()

        for node, parent in self.fetch_nodes_for_root_node(self.root_node):
            pointer = GraphPointer(node, parent)
            for resolvedPointer in self.resolve_pointer(pointer):
                pointers.add(resolvedPointer)
        return pointers

    def set_valid(self, pointers: Pointers, valid: bool):
        for pointer in pointers:
            pointer.valid = valid

    def parse(self, current_pointers: Pointers, codePoint: int) -> Pointers:
        for rule, rule_pointers in self.iterate_over_pointers(current_pointers):
            if is_rule_char(rule):
                valid = False
                for possible_code_point in rule.value:
                    if valid is True:
                        pass
                    elif is_range(possible_code_point):
                        if is_point_in_range(codePoint, possible_code_point):
                            valid = True
                        else:
                            valid = False
                    elif codePoint == possible_code_point:
                        valid = True
                self.set_valid(rule_pointers, valid)

            elif is_rule_char_exclude(rule):
                valid = True
                for possible_code_point in rule.value:
                    if valid is False:
                        pass
                    elif is_range(possible_code_point):
                        if is_point_in_range(codePoint, possible_code_point):
                            valid = False
                        else:
                            valid = True

                self.set_valid(rule_pointers, valid)

            elif not is_rule_end(rule):
                raise ValueError(f"Unsupported rule: {rule}")

        # a pointer's id is the sum of its node's id and its parent's id chain.
        # if two pointers share the same id, it means they point to the same node and have identical parent chains.
        # for the purposes of walking the graph, we only need to keep one of them.
        next_pointers = make_pointers()
        for current_pointer in current_pointers:
            for unresolved_next_pointer in current_pointer.fetch_next():
                for resolved_next_pointer in self.resolve_pointer(
                    unresolved_next_pointer,
                ):
                    next_pointers.add(resolved_next_pointer)
        return next_pointers

    def resolve_pointer(
        self,
        unresolved_pointer: GraphPointer,
    ) -> Iterable[ResolvedGraphPointer]:
        for resolved_pointer in unresolved_pointer.resolve():
            if is_rule_ref(resolved_pointer.node.rule):
                raise ValueError(
                    "Encountered a reference rule when building pointers to the graph",
                )
            if (
                is_rule_end(resolved_pointer.node.rule)
                and resolved_pointer.parent is not None
            ):
                raise ValueError(
                    "Encountered an ending rule with a parent when building pointers to the graph",
                )
            yield resolved_pointer

    def add(self, src: ValidInput, _pointers: Pointers | None = None) -> Pointers:
        pointers = _pointers or self.get_initial_pointers()

        code_points = get_input_as_code_points(src)

        for code_point_pos in range(len(code_points)):
            code_point = code_points[code_point_pos]
            pointers = self.parse(pointers, code_point)
            if len(pointers) == 0:
                raise InputParseError(
                    code_points,
                    code_point_pos,
                    self.previous_code_points,
                )
        self.previous_code_points.extend(code_points)
        return pointers

    #   // generator that yields either the node, or if a reference rule, the referenced node
    #   // we need these function, as distinct from leveraging the logic in GraphPointer,
    #   // because that needs a rule ref with already defined nodes; this function is used to _set_ those nodes
    def fetch_nodes_for_root_node(
        self,
        root_nodes: dict[int, GraphNode],
        parent: GraphPointer | None = None,
    ) -> Iterable[tuple[GraphNode, GraphPointer | None]]:
        for node in root_nodes.values():
            if is_rule_ref(node.rule):
                yield from self.fetch_nodes_for_root_node(
                    self.get_root_node(node.rule.value),
                    GraphPointer(node, parent),
                )
            else:
                yield node, parent

    def print(self, pointers: Pointers | None = None, colors: bool = False) -> str:
        nodes: list[list[GraphNode]] = [
            list(root_node.values()) for root_node in self.roots.values()
        ]
        graph_view: list[str] = []
        for root_node in nodes:
            for node in root_node:
                graph_view.extend(
                    [
                        node.print(
                            {
                                "pointers": pointers or set(),
                                "show_position": True,
                                "colorize": colorize if colors else lambda s, _: str(s),
                            },
                        ),
                    ],
                )

        return "\n".join(graph_view)

    def iterate_over_pointers(
        self,
        pointers: Pointers,
    ) -> Iterable[tuple[UnresolvedRule, Pointers]]:
        seen_rules: dict[UnresolvedRule, Pointers] = {}
        for pointer in pointers:
            rule = pointer.rule
            if is_rule_ref(rule):
                raise ValueError("Encountered a reference rule in the graph")
            seen_rule = seen_rules.get(rule)
            if seen_rule is None:
                seen_rule = set([pointer])
                seen_rules[rule] = seen_rule
            else:
                seen_rule.add(pointer)
        yield from seen_rules.items()

File: packages/gbnf/javascript/src/grammar-graph/graph.ts
import { GraphPointer, } from "./graph-pointer.js";
import { GraphNode, } from "./graph-node.js";
import { getSerializedRuleKey, } from "./get-serialized-rule-key.js";
import { colorize, } from "./colorize.js";
import { GenericSet, } from "./generic-set.js";
import {
  type UnresolvedRule,
  type Pointers,
  customInspectSymbol,
  type ValidInput,
  type ResolvedGraphPointer,
} from "./types.js";
import {
  isRange,
  isRuleChar,
  isRuleCharExcluded,
  isRuleEnd,
  isRuleRef,
} from './type-guards.js';
import { isPointInRange, } from "../utils/is-point-in-range.js";
import {
  InputParseError,
} from "../utils/errors/input-parse-error.js";
import { RuleRef, } from "./rule-ref.js";
import { getInputAsCodePoints, } from "./get-input-as-code-points.js";

type RootNode = Map<number, GraphNode>;
const makePointers = () => new GenericSet<ResolvedGraphPointer, string>(p => p.id);
export class Graph {
  private roots = new Map<number, RootNode>();
  #rootNode?: RootNode;
  grammar: string;
  private previousCodePoints: number[] = [];

  constructor(grammar: string, stackedRules: UnresolvedRule[][][], rootId: number) {
    this.grammar = grammar;
    const ruleRefs: RuleRef[] = [];
    const uniqueRules = new GenericSet<UnresolvedRule, string>(rule => getSerializedRuleKey(rule));
    for (let stackId = 0; stackId < stackedRules.length; stackId++) {
      const stack = stackedRules[stackId];
      const nodes = new Map<number, GraphNode>();
      for (let pathId = 0; pathId < stack.length; pathId++) {
        const path = stack[pathId];
        let node: GraphNode | undefined = undefined;
        for (let stepId = path.length - 1; stepId >= 0; stepId--) {
          const next: undefined | GraphNode = node;
          const rule = stack[pathId][stepId];
          uniqueRules.add(rule);
          if (isRuleRef(rule)) {
            ruleRefs.push(rule);
          }
          // rules coming in may be identical but have different references.
          // here, we ensure we always use the same reference for an identical rule.
          // this makes future comparisons easier.
          const uniqueRule = uniqueRules.get(rule);
          if (uniqueRule === undefined) {
            throw new Error('Could not get unique rule');
          }
          node = new GraphNode(uniqueRule, { stackId, pathId, stepId, }, next);
        }
        if (node === undefined) {
          throw new Error('Could not get node');
        }
        nodes.set(pathId, node);
      }

      this.roots.set(stackId, nodes);
    }
    const __rootNode = this.roots.get(rootId);
    if (!__rootNode) {
      throw new Error(`Root node not found for value: ${rootId}`);
    }
    this.#rootNode = __rootNode;

    for (const ruleRef of ruleRefs) {
      const referencedNodes = new Set<GraphNode>();
      for (const node of this.getRootNode(ruleRef.value).values()) {
        referencedNodes.add(node);
      }
      ruleRef.nodes = referencedNodes;
    }
  }

  private getRootNode = (value: number): RootNode => {
    const rootNode = this.roots.get(value);
    if (!rootNode) {
      throw new Error(`Root node not found for value: ${value}`);
    }
    return rootNode;
  };


  private getInitialPointers = (): Pointers => {
    const pointers = makePointers();

    const rootNode = this.#rootNode;
    if (!rootNode) {
      throw new Error('Root node is not defined');
    }

    for (const { node, parent, } of this.fetchNodesForRootNode(rootNode)) {
      const pointer = new GraphPointer(node, parent);
      for (const resolvedPointer of this.resolvePointer(pointer)) {
        pointers.add(resolvedPointer);
      }
    }
    return pointers;
  };

  private setValid(pointers: GraphPointer[], valid: boolean) {
    for (const pointer of pointers) {
      pointer.valid = valid;
    }
  }

  private parse(currentPointers: Pointers, codePoint: number): Pointers {
    for (const { rule, graphPointers, } of this.iterateOverPointers(currentPointers)) {
      if (isRuleChar(rule)) {
        const valid = rule.value.reduce((
          isValid,
          possibleCodePoint,
        ) => {
          if (isValid) {
            return true;
          }
          if (isRange(possibleCodePoint)) {
            return isPointInRange(codePoint, possibleCodePoint);
          }
          return codePoint === possibleCodePoint;
        }, false);
        this.setValid(graphPointers, valid);
      } else if (isRuleCharExcluded(rule)) {
        const valid = rule.value.reduce((
          isValid,
          possibleCodePoint,
        ) => {
          if (!isValid) {
            return false;
          }
          return isRange(possibleCodePoint) ? !isPointInRange(codePoint, possibleCodePoint) : codePoint !== possibleCodePoint;
        }, true);
        this.setValid(graphPointers, valid);
      } else if (!isRuleEnd(rule)) {
        throw new Error(`Unsupported rule: ${JSON.stringify(rule)}`);
      }
    }

    // a pointer's id is the sum of its node's id and its parent's id chain.
    // if two pointers share the same id, it means they point to the same node and have identical parent chains.
    // for the purposes of walking the graph, we only need to keep one of them.
    const nextPointers = makePointers();
    for (const currentPointer of currentPointers) {
      for (const unresolvedNextPointer of currentPointer.fetchNext()) {
        for (const resolvedNextPointer of this.resolvePointer(unresolvedNextPointer)) {
          nextPointers.add(resolvedNextPointer);
        }
      }
    }
    return nextPointers;
  }

  private * resolvePointer(unresolvedPointer: GraphPointer): IterableIterator<ResolvedGraphPointer> {
    for (const resolvedPointer of unresolvedPointer.resolve()) {
      if (isRuleRef(resolvedPointer.node.rule)) {
        throw new Error('Encountered a reference rule when building pointers to the graph');
      }
      if (isRuleEnd(resolvedPointer.node.rule) && !!resolvedPointer.parent) {
        throw new Error('Encountered an ending rule with a parent when building pointers to the graph');
      }
      yield resolvedPointer;
    }
  }

  public add = (src: ValidInput, _pointers?: Pointers,): Pointers => {
    let pointers = _pointers || this.getInitialPointers();
    const codePoints = getInputAsCodePoints(src);
    for (let codePointPos = 0; codePointPos < codePoints.length; codePointPos++) {
      const codePoint = codePoints[codePointPos];
      pointers = this.parse(pointers, codePoint);
      if (pointers.size === 0) {
        throw new InputParseError(codePoints, codePointPos, this.previousCodePoints);
      }
    }
    this.previousCodePoints.push(...codePoints);
    return pointers;
  };

  // generator that yields either the node, or if a reference rule, the referenced node
  // we need these function, as distinct from leveraging the logic in GraphPointer,
  // because that needs a rule ref with already defined nodes; this function is used to _set_ those nodes
  private * fetchNodesForRootNode(
    rootNodes: Map<number, GraphNode>,
    parent?: GraphPointer,
  ): IterableIterator<{ node: GraphNode; parent?: GraphPointer; }> {
    for (const node of rootNodes.values()) {
      if (isRuleRef(node.rule)) {
        yield* this.fetchNodesForRootNode(this.getRootNode(node.rule.value), new GraphPointer(node, parent));
      } else {
        yield { node, parent, };
      }
    }
  }


  [customInspectSymbol](
    // depth: number, inspectOptions: InspectOptions, inspect: CustomInspectFunction
  ) {
    return this.print({ colors: true, });
  }

  public print = ({ pointers, colors = false, }: { pointers?: Pointers; colors?: boolean } = {}) => {
    const nodes: GraphNode[][] = Array.from(this.roots.values()).map(nodes => Array.from(nodes.values()));
    const graphView = nodes.reduce<string[]>((acc, rootNode) => acc.concat(rootNode.map(node => node.print({
      pointers,
      showPosition: true,
      colorize: colors ? colorize : str => `${str}`,
    }))), []);
    return `\n${graphView.join('\n')}`;
  };

  private * iterateOverPointers(pointers: Pointers): IterableIterator<{ rule: UnresolvedRule; graphPointers: GraphPointer[]; }> {
    const seenRules = new Map<UnresolvedRule, GraphPointer[]>();
    for (const pointer of pointers) {
      const rule = pointer.rule;
      if (isRuleRef(rule)) {
        throw new Error('Encountered a reference rule in the graph, this should not happen');
      }
      let seenRule = seenRules.get(rule);
      if (!seenRule) {
        seenRule = [pointer,];
        seenRules.set(rule, seenRule,);
      } else {
        seenRule.push(pointer);
      }
    }

    for (const [rule, graphPointers,] of seenRules.entries()) {
      yield { rule, graphPointers, };
    }
  }
}
